content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+                                                   +------------+
+```
+
+Then it's a matter of writing some software that runs in the controller and waits for the login prompt.
+
+> You can choose whatever language you feel comfortable with writing tests, but I'd highly suggest looking into either
+Python, Perl or Tcl. Interpreted languages, dynamic are generally better for these tasks because you want to chug out
+tests quickly. Also, it makes sense to use a language that is familiar to all in your team.  
+
+Wiping up some Python code I got this:
+
+``` python
+import unittest
+import serial
+import time
+import xmlrunner
+
+
+class SerialPromptTest(unittest.TestCase):
+    SERIAL_PORT = "/dev/tty.usbserial-110085063"
+    BAUD_RATE = 115200
+    TIMEOUT = 40
+    PROMPT = "verdin-imx8mp-06817296 login:"
+
+    def test_wait_for_PROMPT(self):
+        try:
+            with serial.Serial(
+                self.SERIAL_PORT, self.BAUD_RATE, timeout=self.TIMEOUT
+            ) as ser:
+                time.sleep(1)
+                ser.reset_input_buffer()
+
+                start_time = time.time()
+                received_data = ""
+
+                while time.time() - start_time < self.TIMEOUT:
+                    if ser.in_waiting > 0:
+                        chunk = ser.read(ser.in_waiting).decode(errors="ignore")
+                        print(chunk, end="")
+                        received_data += chunk
+
+                        if self.PROMPT in received_data:
+                            print(f"\nFound expected prompt '{self.PROMPT}'")
+                            return
+                    else:
+                        time.sleep(0.1)
+
+                self.fail(
+                    f"Expected prompt '{self.PROMPT}' not found within {self.TIMEOUT} seconds."
+                )
+
+        except serial.SerialException as e:
+            self.fail(f"Serial connection failed: {e}")
+
+
+if __name__ == "__main__":
+    with open("serial_test_results.xml", "wb") as output:
+        unittest.main(
+            testRunner=xmlrunner.XMLTestRunner(output=output), exit=False
+        )
+```
+
+The `test_wait_for_PROMPT` is quite simple and can be reutilized many times. The basic gist is that we keep looking at
+data chunks coming from the UART, decoding them and waiting until our desired prompt (`verdin-imx8mp-06817296 login:`)
+shows up.
+
+Because this function is encapsulated into a module that calls `unittest.main`, it can output the tests results in a
+standardized format for us, JUnit.xml, which can be interpreted by many other services such as GitLab, GitHub and others.
+
+Another options is using the handy `junit2html` program, which renders the test results as easily digestible HTML pages
+that can be viewed with a web browser.
+
+{{< figure src="/software-testing-part-2-junit2html.png" title="junit2html result of our automated test" alt="" width="50%" class="centered-figure">}}
```

Comment:
Once rendered, this is so small that only ants can see it.

Reply:
I have vision impairment, so I always keep everything like on 200% Zoom. I tried on 100% and it does look small.

Our Hugo instance isn't configured to enlarge images when we display them as well, users can Right Click -> Open in a different tab for now (or ⌘+ or Ctrl+).